lib/update/reconciler.go

```diff
@@ -65,34 +67,31 @@ func (r *reconciler) ReconcilePlan(ctx context.Context, plan storage.OperationPl
 }
 
 func (r *reconciler) trySyncChangelogToEtcd(ctx context.Context) error {
-	shouldSync, err := isEtcdAvailable(ctx, r.FieldLogger)
-	if err != nil {
-		return trace.Wrap(err)
+	disabled, err := isEtcdDisabled(ctx, r.FieldLogger)
+	if err == nil && disabled {
+		r.Info("Etcd disabled, skipping plan sync.")
+		return nil
 	}
-
-	if shouldSync {
-		return trace.Wrap(r.syncChangelog(r.localBackend, r.backend))
-	}
-
-	return nil
+	return trace.Wrap(r.syncChangelog(ctx, r.localBackend, r.backend))
 }
 
 func (r *reconciler) trySyncChangelogFromEtcd(ctx context.Context) error {
-	shouldSync, err := isEtcdAvailable(ctx, r.FieldLogger)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	if shouldSync {
-		return trace.Wrap(r.syncChangelog(r.backend, r.localBackend))
+	disabled, err := isEtcdDisabled(ctx, r.FieldLogger)
+	if err == nil && disabled {
+		r.Info("Etcd disabled, skipping plan sync.")
+		return nil
 	}
-
-	return nil
+	return trace.Wrap(r.syncChangelog(ctx, r.backend, r.localBackend))
 }
 
 // syncChangelog will sync changelog entries from src to dst storage
-func (r *reconciler) syncChangelog(src storage.Backend, dst storage.Backend) error {
-	return trace.Wrap(SyncChangelog(src, dst, r.cluster, r.operationID))
+func (r *reconciler) syncChangelog(ctx context.Context, src storage.Backend, dst storage.Backend) error {
```

Comment:
I'm wondering if retrying reconciliation would actually help. I don't recall seeing reconciliation errors in the logs so I wonder if the problem may actually be a lever deeper, with etcd itself returning stale data due to its leader election and other timeouts? Maybe it would be better to retry the pre-requisites check itself several times to account for situations where etcd needs some time to figure out its leader elections and start returning correct data under load?

Reply:
Not sure how retrying the pre-requisites check can help mitigate a stale read? If we're seeing a stale value from etcd, then we will not be able to tell whether it is stale or up-to-date unless we request the read to be from the quorum.